### ISSUE TITLE
podman exposes udp port on tcp-only image

### ISSUE THREAD

**Issue by edsantiago** [projectatomic/libpod#735]
podman with `--publish-all` exposes udp even if image is tcp-only:
```
# cat >Dockerfile
FROM centos
EXPOSE 8080      ! default is tcp, but you can append '/tcp' if you like; no diff
^D
# img=$(podman build --quiet .)
# cid=$(podman run -d --publish-all $img sleep 60)
# podman port $cid
8080/udp -> 0.0.0.0:35365
8080/tcp -> 0.0.0.0:35365
```
It doesn't seem to be buildah's fault:
```
# podman inspect $img|grep -2 8080
        "ContainerConfig": {
            "ExposedPorts": {
                "8080/tcp": {}
            },
            "Cmd": [
```
RPMs:
```
buildah-0.16-1524754276.git83f4830.fc27.x86_64
podman-0.5.2-1525700809.gitfaa8c3e.fc27.x86_64
```

**Comment by baude** [comment]
We talked a little about this on IRC.  @rhatdan , if you recall, when we started enabling the network stack I asked about whether we need to distinguish between the tcp and udp protocols.  At that time, we jointly made the decision to treat them both as one.  This is obviously a departure from what other container runtimes do.

Fixing the example that @edsantiago provides may actually be doable ... but if this has to be be extended into -p and friends, it will be a significant amount of work.

**Comment by rhatdan** [comment]
@edsantiago WDYT?  Is this a real issue?

**Comment by edsantiago** [comment]
The worst (and least unlikely) scenario I can think of where this matters is 53: I can envision a container running BIND, where the creator wants to expose 53/udp but does not want to allow AXFR on 53/tcp, and they do so by filtering instead of by configuration. It would tick me off if I were to discover that my container was not running as I had crafted it. But I'm not the sort of person who would do that.

IMO this isn't worth five days' effort but if it's easy to do a subset, it's worth considering.

Disclaimer: there are almost certainly other cases I haven't thought of.

**Comment by rhatdan** [comment]
Ok lets leave this as an issue, but put it to the bottom of the queue.  Or if some contributor wants to take it on, go for it.

**Comment by rhatdan** [comment]
@baude Is this something we should have an intern look at?

**Comment by baude** [comment]
I'd prefer to fix it given the history of this code ... the larger question is if we are going to still assume -P 80 for example is both udp and tcp.  and if we are going to start respecting -P 80/tcp

**Comment by rhatdan** [comment]
Yes I think -P 80 means both and then you have to give specific for tcp and udp.

**Comment by kunalkushwaha** [comment]
this issue seems still persist. @baude if you are busy, may I try to fix this?

```
$ cat Dockerfile.builah 
FROM docker.io/library/centos:latest
EXPOSE 8080

$ cid=$(sudo podman run -d -p 6061/tcp $img sleep 60)

$ sudo podman port $cid                              
6061/udp -> 0.0.0.0:34753
6061/tcp -> 0.0.0.0:34753

```

**Comment by mheon** [comment]
@kunalkushwaha I don't think anyone is actively working on this issue, so feel free

**Comment by baude** [comment]
Yes, feel free

**Comment by rhatdan** [comment]
@kunalkushwaha We would love to have you work on this.  Thanks.

**Comment by kunalkushwaha** [comment]
Thanks, I will start looking into this.

**Comment by kunalkushwaha** [comment]
Just to review, I have a put down the current and expected behaviour of  `-p` and `-P|--publish-all` output of ports. 

|   Options	|  Now  	|  Expected 	| 
|---	|---	|---	|
| `-p 80`   	|   `80/tcp`, `80/udp`	|  `80/tcp`  	|
| `-p 80/tcp`  	|  `80/tcp`, `80/udp` 	| `80/tcp`  	|
|   `-p 80/udp`	| `80/tcp`, `80/udp`  	|  `80/udp` 	|
|   `-P` with `80` in image 	| `80/tcp`, `80/udp`  	| `80/tcp`, `80/udp`*	|
|   `-P` with `80/udp` in image	| `80/tcp`, `80/udp`  	|  `80/udp` 	|

*_In other runtimes like docker ,here `80/tcp` is exposed_

Please let me  know, if any changes in expected behaviour is required.

**Comment by rhatdan** [comment]
LGTM

**Comment by baude** [comment]
When the protocol is not stated, are both udp and tcp ports supposed to be used?

**Comment by mheon** [comment]
Per docker it's just TCP if not specified.

**Comment by baude** [comment]
ok, works for me!

**Comment by edsantiago** [comment]
Reference: [EXPOSE](https://docs.docker.com/engine/reference/builder/#expose) is documented as explicitly defaulting to tcp. To my surprise, I've been unable to find a definitive reference for `-p`.

**Comment by kunalkushwaha** [comment]
While debugging, I found, in case of no proto defined for exposed ports in Image(Dockerfile), the `tcp` is value we get after Image is parsed as per  `ociV1 Spec`
- the `tcp`  value is returned from function [`ociv1Image()`](https://github.com/containers/libpod/blob/master/libpod/image/image.go#L787) 

Also I cross checked the [OCI specs](https://github.com/opencontainers/image-spec/blob/master/config.md#properties), and it states
 > __ExposedPorts object, OPTIONAL__
> A set of ports to expose from a container running this image. Its keys can be in the format of: port/tcp, port/udp, port with __the default protocol being tcp if not specified.__

I feel we should go as per specs i.e. default `tcp` port exposed in case of no proto  specified.
So updated table of current vs expected will be 

|   Options	|  Now  	|  Expected 	| 
|---	|---	|---	|
| `-p 80`   	|   `80/tcp`, `80/udp`	|  `80/tcp`  	|
| `-p 80/tcp`  	|  `80/tcp`, `80/udp` 	| `80/tcp`  	|
|   `-p 80/udp`	| `80/tcp`, `80/udp`  	|  `80/udp` 	|
|   `-P` with `80` in image 	| `80/tcp`, `80/udp`  	| `80/tcp`|
|   `-P` with `80/udp` in image	| `80/tcp`, `80/udp`  	|  `80/udp` 	|

Please let me know WDYT

**Comment by rhatdan** [comment]
SGTM

**Comment by kunalkushwaha** [comment]
This issue can be closed now.
